ingestion_server/ingestion_server/api.py

```diff
@@ -47,124 +47,175 @@ def _get_base_url(req):
+            res.media = {
                 "message": "Failed to schedule task due to an internal server "
                 "error. Check scheduler logs."
             }
-            return
 
-    def on_get(self, req, resp):
-        """List all indexing tasks."""
-        resp.media = self.tracker.list_task_statuses()
+    def on_get(self, _, res):
+        """
+        Handles an incoming GET request. Provides information about all past tasks.
+
+        :param _: the incoming request
+        :param res: the appropriate response
+        """
+
+        res.media = self.tracker.list_task_statuses()
 
 
 class TaskStatus(BaseTaskResource):
-    def on_get(self, req, resp, task_id):
-        """Check the status of a single task."""
-        task = self.tracker.id_task.get(task_id)
-        if task is None:
-            resp.status = falcon.HTTP_404
-            resp.media = {"message": f"No task found with id {task_id}"}
-            return
-
-        percent_completed = self.tracker.id_progress[task_id].value
-        active_workers = bool(self.tracker.id_active_workers[task_id].value)
-        active = task.is_alive() or active_workers
-
-        resp.media = {
-            "active": active,
-            "percent_completed": percent_completed,
-            "error": percent_completed < 100 and not active,
-        }
+    def on_get(self, _, res, task_id):
+        """
+        Handles an incoming GET request. Provides information about a single task.
+
+        :param _: the incoming request
+        :param res: the appropriate response
+        :param task_id: the ID of the task for which to get the information
+        """
 
+        try:
+            result = self.tracker.get_task_status(task_id)
+            res.media = result
+        except KeyError:
+            res.status = falcon.HTTP_404
+            res.media = {"message": f"No task found with id {task_id}."}
 
-class WorkerFinishedResource(BaseTaskResource):
-    """
-    For notifying ingestion server that an indexing worker has finished its
-    task.
-    """
 
+class WorkerFinishedResource(BaseTaskResource):
     def on_post(self, req, _):
+        """
+        Handles an incoming POST request. Records messages sent from indexer workers.
+
+        :param req: the incoming request
+        :param _: the appropriate response
+        """
+
         task_data = worker_finished(str(req.remote_addr), req.media["error"])
         task_id = task_data.task_id
         target_index = task_data.target_index
-        active_workers = self.tracker.id_active_workers[task_id]
+        task_info = self.tracker.tasks[task_id]
+        active_workers = task_info["active_workers"]
 
         # Update global task progress based on worker results
-        self.tracker.id_progress[task_id].value = task_data.percent_successful
+        task_info["progress"].value = task_data.percent_successful
 
         if task_data.percent_successful == 100:
             logging.info(
```

Comment:
The log below (`Attempting to promote index {target_index}`) is no longer accurate, now that the promotion step has been moved out of `ingest_upstream`.